custom script from git component finalised before review (#1005)
appcircleio/appcircle-docusaurus: docs/workflows/common-workflow-steps/custom-script-from-git.md

---
title: Custom Script from Git
description: Use Script steps for additional functionalities in your builds.

> tags: [custom scripts, build, test, workflow, step]

Reviewer: The tags "test" do not suit this component; "workflow" and "step" are too general for this component.
More relevant tags to add would be "git" and "git scripts" to better reflect the feature’s focus on running scripts from Git repositories.

---

import Tabs from '@theme/Tabs';

> import TabItem from '@theme/TabItem';

Reviewer: Where do you use those libraries?

import Screenshot from '@site/src/components/Screenshot';
import SensitiveVariablesDanger from '@site/docs/\_sensitive-variables-danger.mdx';

# Custom Script from Git

You can use **Custom Script from Git** to clone and run your own scripts directly from a Git repository as part of your Appcircle build. This step supports authenticated cloning (via username and PAT).

Before execution, the step will clone (or reuse) your repository, check out the specified branch, and execute the script based on its file extension (postfix).

:::caution

Keep in mind that the script file has to be under the repository root folder in order to work with the component.

> :::

Reviewer: Can we talk about this issue?

:::info

> The step inspects your script file’s postfix (its extension—.sh, .py, .rb, .pl, .js, .java) to choose the execution path. It must match exactly.

Reviewer: Please use backticks `` to format technical terms and code elements for better readability and clarity.

Supported script types and their execution commands:

- example-script.sh → bash
- example-script.rb → ruby
- example-script.py → python
- example-script.pl → perl
- example-script.js → nodejs

> - example-script.java →

Reviewer: Please provide example file names that follow the naming conventions for each programming language. For instance:

```
example_script.rb → Ruby
ExampleScript.java → Java
```

Source for Java: https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

```ruby 

> # Script-filename and Class-name within the script must match.

Reviewer: Could you please move this warning outside the code block? As a comment, it is less visible, but this warning seems important and should stand out more clearly.

javac example-script.java
java example-script 
```

:::

:::caution

If you are seeing the following error in the build log, please ensure that both the username and the personal access token (PAT) are set correctly. This error is returned by the Git provider when authentication fails due to an incorrect or missing username or PAT:

```bash

fatal: could not read Username for Git provider

```

:::

### Prerequisites

There are no prerequisites required before using the **Custom Script from Git** step.

:::tip

Note that you can put the **Custom Script from Git** component anywhere you want in the workflow. This step is used to add different capabilities to the existing workflow. As well as the [**Custom Script**](/workflows/common-workflow-steps/custom-script).

:::

<Screenshot url='https://cdn.appcircle.io/docs/assets/BE6419-csfromgit.png' />

### Input Variables

This step contains some input variable(s). It needs these variable(s) to work. The table below gives explanation for this variable(s).

<Screenshot url='https://cdn.appcircle.io/docs/assets/BE6419-csfromgitinput.png' />

> <SensitiveVariablesDanger />

Reviewer: To help users compare the screenshot with the table more easily, could you please place this warning on top of the image instead of below or inside the code? This positioning will make the warning more visible and useful.

| Variable Name                  | Description                                                                                                                                                                                       | Status   |
|--------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|

> | `$AC_SCRIPT_FILENAME`          | Specifies the exact name of the script file to execute within the Git repository (e.g., test.sh, code.rb, python.py).                                                                             | Required |

Reviewer: ```suggestion
| `$AC_SCRIPT_FILENAME`          | Specifies the exact name of the script file to execute within the Git repository (e.g., `test.sh`, `code.rb`, `python.py`).                                                                             | Required |
```

> | `$AC_SCRIPT_EXTRA_PARAMETERS`  | Additional parameters to pass to the relevant script file. Separate multiple parameters with a comma \" , \". If a parameter contains spaces, enclose it in \" \".                                | Optional |

Reviewer: CAn you add an example

> | `$AC_SCRIPT_REPO_DIR`          | If the Git repository has already been cloned in a previous workflow step, set Repository Directory Output here. **Required** if Clone URL, Username, and Personal Access Token are not provided. | Optional |

Reviewer: ```suggestion
| `$AC_SCRIPT_REPO_DIR`          | If the Git repository has already been cloned in a previous workflow step, set the repository directory output here. This input is **required** if the Git clone URL is not provided. | Optional |
```

| `$AC_SCRIPT_REPO_CLONE_URL`    | Git repository clone URL (https). **Required** if the Repository Directory Output is not provided.                                                                                                | Optional |

> | `$AC_SCRIPT_GIT_USERNAME`      | Git provider username for authentication. **Required** if the Repository Directory Output is not provided. **Public repository** doesnt require Username.                                         | Optional |

Reviewer: ```suggestion
| `$AC_SCRIPT_GIT_USERNAME`      | Git provider username for authentication. **Required** if the repository directory output is not provided and the Git clone URL points to a private repository.                                         | Optional |
```

> | `$AC_SCRIPT_GIT_PAT`           | Git provider personal access token for authentication. **Required** if the Repository Directory Output is not provided. **Public repository** doesnt require PAT.                                 | Optional |

Reviewer: ```suggestion
| `$AC_SCRIPT_GIT_PAT`           | Git provider personal access token for authentication. **Required** if the repository directory output is not provided and the Git clone URL points to a private repository.                                 | Optional |
```
